src/MongoDB.Driver/ClusterKey.cs

```diff
@@ -34,13 +35,15 @@ internal class ClusterKey
         private readonly TimeSpan _heartbeatInterval;
         private readonly TimeSpan _heartbeatTimeout;
         private readonly bool _ipv6;
+        private readonly IReadOnlyDictionary<string, IReadOnlyDictionary<string, object>> _kmsProviders;
```

Comment:
@rstam can you elaborate on what's troubling to you about them?  Are the equality semantics ill-defined?

Reply:
Actually I forgot to even review the implementation of Equals... although that looks like it was more difficult than expected.

What bothered me was whether differences in these values were a proper reason to have to create a new cluster.

Do we really want to be creating a new cluster just because the encryption settings differ?

Reply:
I agree that it's awkward. But if CryptClient is part of the Cluster what choice do we have?

Reply:
We have the choice to reexamine whether this was the right choice or not.

When Vincent asked why CryptClient wasn't owned by the Cluster and I proposed that it was worth looking into I did not expect that it would result in possibly creating unnecessary Clusters.

Unnecessary Clusters are a VERY BAD THING, because they never go away, and each Cluster has a connection pool so extra Clusters mean lots of extra sockets.

Reply:
I think we have two not-so-good alternatives:

- Have the MongoClient own it and risk re-creating it unnecessarily.  The mitigating factor here is that we could tell people not to create more MongoClient instances than they need.  Robert, I think you suggested this before
- Have the Cluster own it and risk otherwise unnecessary Cluster instances.  The mitigating factor here is that I wouldn't expect more than two for the same cluster.  One with a CryptClient and one without, and in most circumstances I'd expect only one.

I don't think there is an obvious right answer here, so someone has to make the call.

Reply:
There is a third alternative I suggested when we were first considering moving `CryptClient` out of `MongoClient`.

We could have a `CryptClientRegistry` in High.

I initially suggested that as a way of keeping `CryptClient` out of `Core`, but it could also be used as a way to keep from creating unnecessary `Clusters`.

I would also be OK with leaving as it is and assuming that applications won't use lots of different `kmsProvider` configurations (or warning them not to in the documentation).

Leaving it the way it is for now that does not preclude using a `CryptClientRegistry` in the furture.